pkg/service/token/token-service.go

```diff
@@ -339,6 +339,87 @@ func (t *tokenService) Revoke(ctx context.Context, rq *connect.Request[apiv2.Tok
 	return connect.NewResponse(&apiv2.TokenServiceRevokeResponse{}), nil
 }
 
+func (t *tokenService) Refresh(ctx context.Context, _ *connect.Request[apiv2.TokenServiceRefreshRequest]) (*connect.Response[apiv2.TokenServiceRefreshResponse], error) {
+	token, ok := tokenutil.TokenFromContext(ctx)
+	if !ok || token == nil {
+		return nil, connect.NewError(connect.CodeUnauthenticated, fmt.Errorf("no token found in request"))
+	}
+
+	oldtoken, err := t.tokens.Get(ctx, token.UserId, token.Uuid)
+	if err != nil {
+		if errors.Is(err, tokenutil.ErrTokenNotFound) {
+			return nil, connect.NewError(connect.CodeNotFound, errors.New("token not found"))
+		}
+		return nil, connect.NewError(connect.CodeInternal, err)
+	}
+
+	// we first copy the token permission from the old token
+	createRequest := &apiv2.TokenServiceCreateRequest{
+		Permissions:  oldtoken.Permissions,
+		ProjectRoles: oldtoken.ProjectRoles,
+		TenantRoles:  oldtoken.TenantRoles,
+		AdminRole:    oldtoken.AdminRole,
+	}
+
+	err = validateTokenCreate(token, createRequest, t.servicePermissions, t.adminSubjects)
+	if err != nil {
+		return nil, connect.NewError(connect.CodePermissionDenied, err)
+	}
```

Comment:
this is just to avoid performing the `projectsAndTenantsGetter` if it already fails here, right?
Because below, there already is another `validateTokenCreate` call.

Reply:
This is true, i think this was introduced by you during the implementation of the token create service :-)

Reply:
I guess I'll have a quick chat about this with @Gerrit91. Maybe he knows why. Otherwise I'd vote for removing one call of it.

Reply:
Gerrit said this is fine

Reply:
Let's keep both validations. The check is not so expensive and it catches things like "if admins have changed over time, prevent token refresh", which is fine to detect here.